Create PermissionsPolicy from Document

szewai opened the pull request:
#### b4bb6216f8304dd9e2bab2648aad261546d4dad5
<pre>
Create PermissionsPolicy from Document
<a href="https://bugs.webkit.org/show_bug.cgi?id=276353">https://bugs.webkit.org/show_bug.cgi?id=276353</a>
<a href="https://rdar.apple.com/131357201">rdar://131357201</a>

Reviewed by Youenn Fablet.

This patch makes a few changes that will help add support for permissions policy with site isolation:
1. Replace BitSet with HashSet for InheritedPolicy in PermissionsPolicy, so that we could use strongly typed enum for
getting and setting, which makes the code easier to understand and it is more safe to use it in IPC later. (We will
need to pass the policy between UI process and web processes).
2. Introduce OwnerPermissionsPolicyData struct, which contains information that Document needs from its owner element to
compute permissions policy. With site isolation, Document may not have access to owner element, so we need to store the
information elsewhere. In this patch, we store it in LocalFrame.
3. Allow PermissionsPolicy to be created from Document, and PermissionsPolicy can decide whether to construct object
from owner element or owner policy.
4. Split Allowlist from PermissionsPolicy so that it&apos;s easier to add IPC serialization format for it later.

The owner policy of LocalFrame is not set yet, so there is no behavior change.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::permissionsPolicy const):
* Source/WebCore/html/Allowlist.h: Added.
(WebCore::Allowlist::Allowlist):
(WebCore::Allowlist::origins const):
(WebCore::Allowlist::matches const):
* Source/WebCore/html/OwnerPermissionsPolicyData.h: Added.
(WebCore::OwnerPermissionsPolicyData::OwnerPermissionsPolicyData):
* Source/WebCore/html/PermissionsPolicy.cpp:
(WebCore::forEachFeature):
(WebCore::parseAllowlist):
(WebCore::PermissionsPolicy::processPermissionsPolicyAttribute):
(WebCore::featureValueForOrigin):
(WebCore::PermissionsPolicy::computeInheritedPolicyValueInContainer const):
(WebCore::PermissionsPolicy::inheritedPolicyValueForFeature const):
(WebCore::PermissionsPolicy::PermissionsPolicy):
(WebCore::PermissionsPolicy::Allowlist::matches const): Deleted.
(WebCore::index): Deleted.
* Source/WebCore/html/PermissionsPolicy.h:
(WebCore::PermissionsPolicy::PermissionsPolicy):
(WebCore::PermissionsPolicy::inheritedPolicy const):
(): Deleted.
(WebCore::PermissionsPolicy::Allowlist::Allowlist): Deleted.
* Source/WebCore/page/LocalFrame.h:

Canonical link: <a href="https://commits.webkit.org/280791@main">https://commits.webkit.org/280791@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0b0909f5a208f4f351f9a387151a76e5d32c1f7f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57602 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36930 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10077 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61224 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8047 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59730 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44566 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8235 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46647 "Passed tests") | [❌ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5718 "1 failures") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59632 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34655 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49776 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27514 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31429 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7068 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7050 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53386 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7341 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62904 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1516 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7430 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53905 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1522 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49789 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54013 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12748 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1294 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32759 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33845 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34929 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33590 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/html/Allowlist.h:
```diff
@@ -0,0 +1,71 @@
+        : m_origins(WTFMove(origins))
+    {
+    }
+    OriginsVariant origins() const { return m_origins; }
```
Do we need a copy or can we just return a const ref?

szewai replied:
I think we can make a const ref.

youennf commented on Source/WebCore/page/LocalFrame.h:
```diff
@@ -318,6 +319,8 @@ class LocalFrame final : public Frame {
     String customUserAgentAsSiteSpecificQuirks() const final;
     String customNavigatorPlatform() const final;
     OptionSet<AdvancedPrivacyProtections> advancedPrivacyProtections() const final;
+    WEBCORE_EXPORT void setOwnerPermissionsPolicy(const OwnerPermissionsPolicyData&);
```
Can it be a &&

szewai replied:
Sure.

youennf commented on Source/WebCore/html/PermissionsPolicy.h:
```diff
@@ -73,28 +73,24 @@ class PermissionsPolicy {
 private:
     bool computeInheritedPolicyValueInContainer(Feature, const HTMLFrameOwnerElement*, const SecurityOriginData&) const;
+    bool computeInheritedPolicyValueInContainer(Feature, const SecurityOriginData&, const PermissionsPolicy&, const PolicyDirective&, const SecurityOriginData&) const;
+    void forEachFeature(const Function<void(Feature)>&);
```
can be static

szewai replied:
Sure.

youennf commented on Source/WebCore/html/Allowlist.h:
```diff
@@ -0,0 +1,71 @@
+public:
+    Allowlist() = default;
+    struct AllowAllOrigins { };
+    explicit Allowlist(AllowAllOrigins allow)
```
We could make it not explicit to just write `return Allowlist::AllowAllOrigins { }`.
Also, should it be named `AllowList` instead of `Allowlist`?

szewai replied:
Sure, will change; I thought about this and found Allowlist is one word according to spec: https://w3c.github.io/webappsec-permissions-policy/#allowlists

youennf commented on Source/WebCore/html/OwnerPermissionsPolicyData.h:
```diff
@@ -0,0 +1,46 @@
+
+struct OwnerPermissionsPolicyData {
+    using PolicyDirective = PermissionsPolicy::PolicyDirective;
+    OwnerPermissionsPolicyData(const SecurityOriginData& documentOrigin, const PermissionsPolicy& documentPolicy, const PolicyDirective& containerPolicy)
```
Maybe we can take &&?

szewai replied:
Sure.

youennf commented on Source/WebCore/html/PermissionsPolicy.cpp:
```diff
@@ -398,50 +389,61 @@ bool PermissionsPolicy::computeInheritedPolicyValueInContainer(Feature feature,
+    std::optional<OwnerPermissionsPolicyData> ownerPolicy;
+    if (RefPtr frame = document.frame())
+        ownerPolicy = frame->ownerPermissionsPolicy();
+    if (ownerPolicy) {
```
ownerPolicy could be defined in ` if (RefPtr frame = document.frame())` block.

szewai replied:
Sure.

szewai committed:
Create PermissionsPolicy from Document
https://bugs.webkit.org/show_bug.cgi?id=276353
rdar://131357201

Reviewed by Youenn Fablet.

This patch makes a few changes that will help add support for permissions policy with site isolation:
1. Replace BitSet with HashSet for InheritedPolicy in PermissionsPolicy, so that we could use strongly typed enum for
getting and setting, which makes the code easier to understand and it is more safe to use it in IPC later. (We will
need to pass the policy between UI process and web processes).
2. Introduce OwnerPermissionsPolicyData struct, which contains information that Document needs from its owner element to
compute permissions policy. With site isolation, Document may not have access to owner element, so we need to store the
information elsewhere. In this patch, we store it in LocalFrame.
3. Allow PermissionsPolicy to be created from Document, and PermissionsPolicy can decide whether to construct object
from owner element or owner policy.
4. Split Allowlist from PermissionsPolicy so that it's easier to add IPC serialization format for it later.

The owner policy of LocalFrame is not set yet, so there is no behavior change.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::permissionsPolicy const):
* Source/WebCore/html/Allowlist.h: Added.
(WebCore::Allowlist::Allowlist):
(WebCore::Allowlist::origins const):
(WebCore::Allowlist::matches const):
* Source/WebCore/html/OwnerPermissionsPolicyData.h: Added.
(WebCore::OwnerPermissionsPolicyData::OwnerPermissionsPolicyData):
* Source/WebCore/html/PermissionsPolicy.cpp:
(WebCore::forEachFeature):
(WebCore::parseAllowlist):
(WebCore::PermissionsPolicy::processPermissionsPolicyAttribute):
(WebCore::featureValueForOrigin):
(WebCore::PermissionsPolicy::computeInheritedPolicyValueInContainer const):
(WebCore::PermissionsPolicy::inheritedPolicyValueForFeature const):
(WebCore::PermissionsPolicy::PermissionsPolicy):
(WebCore::PermissionsPolicy::Allowlist::matches const): Deleted.
(WebCore::index): Deleted.
* Source/WebCore/html/PermissionsPolicy.h:
(WebCore::PermissionsPolicy::PermissionsPolicy):
(WebCore::PermissionsPolicy::inheritedPolicy const):
(): Deleted.
(WebCore::PermissionsPolicy::Allowlist::Allowlist): Deleted.
* Source/WebCore/page/LocalFrame.h:

Canonical link: https://commits.webkit.org/280791@main
